src/init.lua

```diff
@@ -20,55 +21,72 @@ local COLLECTION_TAG = "WindShake" -- The CollectionService tag to be watched an
 
 local FALLBACK_SETTINGS = {
 	WindDirection = Vector3.new(0.5, 0, 0.5),
-	WindSpeed = 20,
 	WindPower = 0.5,
+	WindSpeed = 20,
 }
 
+type Settings = Settings.Class
+
 -----------------------------------------------------------------------------------------------------------------
 
-local ObjectShakeAddedEvent = Instance.new("BindableEvent")
-local ObjectShakeRemovedEvent = Instance.new("BindableEvent")
-local ObjectShakeUpdatedEvent = Instance.new("BindableEvent")
-local PausedEvent = Instance.new("BindableEvent")
-local ResumedEvent = Instance.new("BindableEvent")
+local Paused = Instance.new("BindableEvent")
+local Resumed = Instance.new("BindableEvent")
+local ObjectShakeAdded = Instance.new("BindableEvent")
+local ObjectShakeRemoved = Instance.new("BindableEvent")
+local ObjectShakeUpdated = Instance.new("BindableEvent")
+
 
 local WindShake = {
 	RenderDistance = 150,
 	MaxRefreshRate = 1 / 60,
 	SharedSettings = Settings.new(script),
 
-	ObjectMetadata = {},
-	VectorMap = VectorMap.new(),
+	ObjectMetadata = {} :: {
+		[Instance]: {
+			ChunkKey: Vector3,
+			Settings: Settings,
+			Seed: number,
+			Origin: CFrame,
+			LastUpdate: number,
+		}
+	},
 
+	VectorMap = VectorMap.new(),
 	Handled = 0,
 	Active = 0,
 
-	_partList = table.create(500),
-	_cframeList = table.create(500),
-
-	ObjectShakeAdded = ObjectShakeAddedEvent.Event,
-	ObjectShakeRemoved = ObjectShakeRemovedEvent.Event,
-	ObjectShakeUpdated = ObjectShakeUpdatedEvent.Event,
-	Paused = PausedEvent.Event,
-	Resumed = ResumedEvent.Event,
+	_partList = table.create(500) :: { any }, -- ('any' because Studio and Luau LSP disagree on the type)
+	_cframeList = table.create(500) :: { CFrame },
+
+	ObjectShakeAdded = ObjectShakeAdded.Event,
+	ObjectShakeRemoved = ObjectShakeRemoved.Event,
+	ObjectShakeUpdated = ObjectShakeUpdated.Event,
+	
+	Paused = Paused.Event,
+	Resumed = Resumed.Event,
+
+	Initialized = nil :: boolean?,
+	AddedConnection = nil :: RBXScriptConnection?,
+	UpdateConnection = nil :: RBXScriptConnection?,
+	RemovedConnection = nil :: RBXScriptConnection?,
+	WorkspaceWindConnection = nil :: RBXScriptConnection?,
 }
 
+type WindShake = typeof(WindShake)
+
 export type WindShakeSettings = {
 	WindDirection: Vector3?,
 	WindSpeed: number?,
 	WindPower: number?,
 }
 
-function WindShake:Connect(funcName: string, event: RBXScriptSignal): RBXScriptConnection
-	local callback = self[funcName]
-	assert(typeof(callback) == "function", "Unknown function: " .. funcName)
-
+local function Connect<Args...>(self: WindShake, event: RBXScriptSignal, callback: (self: WindShake, Args...) -> ()): RBXScriptConnection
 	return event:Connect(function(...)
 		return callback(self, ...)
 	end)
 end
 
-function WindShake:AddObjectShake(object: BasePart | Bone, settingsTable: WindShakeSettings?)
+function WindShake.AddObjectShake(self: WindShake, object: Instance, settingsTable: WindShakeSettings?)
```

Comment:
Why the change from `BasePart | Bone`?

Reply:
It's because the function was being directly plugged into the CollectionService callback, so it technically *was* consuming types of `Instance`. I can correct this behavior in the `Init` function.